File: bindings/python/src/rockbox_ffi/dsp.py
"""The DSP pipeline: EQ, tone, surround, compressor, ReplayGain, resampler."""

from __future__ import annotations

import array
import math
from typing import Iterable, Optional

from ._ffi import ffi, lib

_NAN = float("nan")


def _opt(v: Optional[float]) -> float:
    """None -> NaN, the ABI's 'tag absent' sentinel."""
    return _NAN if v is None else float(v)


class Dsp:
    """Interleaved-S16LE-stereo DSP instance.

    The underlying ``dsp_config`` is a process-wide singleton, so only one
    :class:`Dsp` may exist at a time and it must be used from one thread.
    Use as a context manager, or call :meth:`close` when done.
    """

    def __init__(self, sample_rate: int):
        self._p = lib.rb_dsp_new(int(sample_rate))
        if self._p == ffi.NULL:
            raise RuntimeError("rb_dsp_new returned NULL")

    # -- lifecycle --------------------------------------------------------
    def close(self) -> None:
        if getattr(self, "_p", ffi.NULL) != ffi.NULL:
            lib.rb_dsp_free(self._p)
            self._p = ffi.NULL

    def __del__(self):
        self.close()

    def __enter__(self) -> "Dsp":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    # -- configuration ----------------------------------------------------
    def set_input_frequency(self, hz: int) -> None:
        lib.rb_dsp_set_input_frequency(self._p, int(hz))

    def flush(self) -> None:
        lib.rb_dsp_flush(self._p)

    def eq_enable(self, enable: bool) -> None:
        lib.rb_dsp_eq_enable(self._p, bool(enable))

    def set_eq_band(self, band: int, cutoff_hz: int, q: float, gain_db: float) -> None:
        """Configure one EQ band (0..=9). Band 0 low shelf, 9 high shelf."""
        lib.rb_dsp_set_eq_band(self._p, int(band), int(cutoff_hz), float(q), float(gain_db))

    def set_eq_precut(self, db: float) -> None:
        lib.rb_dsp_set_eq_precut(self._p, float(db))

    def set_tone(self, bass_db: int, treble_db: int) -> None:
        lib.rb_dsp_set_tone(self._p, int(bass_db), int(treble_db))

    def set_tone_cutoffs(self, bass_hz: int, treble_hz: int) -> None:
        lib.rb_dsp_set_tone_cutoffs(self._p, int(bass_hz), int(treble_hz))

    def set_surround(self, delay_ms: int, balance: int, fx1: int, fx2: int) -> None:
        lib.rb_dsp_set_surround(self._p, int(delay_ms), int(balance), int(fx1), int(fx2))

    def set_channel_config(self, mode: int) -> None:
        lib.rb_dsp_set_channel_config(self._p, int(mode))

    def set_stereo_width(self, percent: int) -> None:
        lib.rb_dsp_set_stereo_width(self._p, int(percent))

    def set_compressor(
        self,
        threshold: int,
        makeup_gain: int,
        ratio: int,
        knee: int,
        release_time: int,
        attack_time: int,
    ) -> None:
        lib.rb_dsp_set_compressor(
            self._p, int(threshold), int(makeup_gain), int(ratio),
            int(knee), int(release_time), int(attack_time),
        )

    def set_replaygain(self, mode: int, noclip: bool, preamp_db: float) -> None:
        """mode: see :class:`rockbox_ffi.enums.DspReplayGainMode`
        (TRACK=0, ALBUM=1, SHUFFLE=2, OFF=3)."""
        lib.rb_dsp_set_replaygain(self._p, int(mode), bool(noclip), float(preamp_db))

    def set_replaygain_gains(
        self,
        track_gain_db: Optional[float] = None,
        album_gain_db: Optional[float] = None,
        track_peak: Optional[float] = None,
        album_peak: Optional[float] = None,
    ) -> None:
        """Per-track gains in plain dB / peaks as linear amplitude
        (1.0 = full scale). None for any absent tag."""
        lib.rb_dsp_set_replaygain_gains(
            self._p, _opt(track_gain_db), _opt(album_gain_db),
            _opt(track_peak), _opt(album_peak),
        )

    def set_replaygain_gains_raw(
        self, track_gain: int, album_gain: int, track_peak: int, album_peak: int
    ) -> None:
        """Native Q7.24 linear factors (the ``raw_*`` fields from
        :func:`rockbox_ffi.metadata.read`), 0 = not tagged."""
        lib.rb_dsp_set_replaygain_gains_raw(
            self._p, int(track_gain), int(album_gain), int(track_peak), int(album_peak)
        )

    # -- processing -------------------------------------------------------
    def process(self, samples: Iterable[int]) -> array.array:
        """Run interleaved stereo S16 ``samples`` through the pipeline.

        Accepts any int16 sequence (list, ``array('h')``, bytes-like via
        ``array``). Returns a new ``array('h')`` of processed samples
        (length may differ from the input — the resampler buffers).
        """
        buf = samples if isinstance(samples, array.array) and samples.typecode == "h" \
            else array.array("h", samples)
        n = len(buf)
        if n % 2 != 0:
            raise ValueError("input must be interleaved stereo (even length)")

        cin = ffi.cast("const int16_t *", ffi.from_buffer(buf))
        out_len = ffi.new("size_t *")
        out_ptr = lib.rb_dsp_process(self._p, cin, n, out_len)

        produced = int(out_len[0])
        if out_ptr == ffi.NULL or produced == 0:
            return array.array("h")
        try:
            result = array.array("h")
            result.frombytes(ffi.buffer(out_ptr, produced * 2)[:])
            return result
        finally:
            lib.rb_buffer_free(out_ptr, produced)


def sine_stereo(freq_hz: float, seconds: float, rate: int, amplitude: int = 16000) -> array.array:
    """Helper: generate ``seconds`` of a sine as interleaved stereo int16."""
    n = int(seconds * rate)
    buf = array.array("h")
    for i in range(n):
        s = int(math.sin(i * 2.0 * math.pi * freq_hz / rate) * amplitude)
        s = max(-32768, min(32767, s))
        buf.append(s)
        buf.append(s)
    return buf

File: bindings/ruby/lib/rockbox_ffi/dsp.rb
# frozen_string_literal: true

module RockboxFFI
  # The DSP pipeline: EQ, tone, surround, compressor, ReplayGain, resampler.
  #
  # The underlying dsp_config is a process-wide singleton, so only one Dsp may
  # exist at a time and it must be used from one thread. Call #close when done,
  # or use the block form Dsp.open(rate) { |dsp| ... }.
  class Dsp
    # Open a Dsp; if a block is given, close it automatically afterwards.
    def self.open(sample_rate)
      dsp = new(sample_rate)
      return dsp unless block_given?

      begin
        yield dsp
      ensure
        dsp.close
      end
    end

    def initialize(sample_rate)
      @ptr = Lib.rb_dsp_new(Integer(sample_rate))
      raise "rb_dsp_new returned NULL" if @ptr.null?

      @finalizer = self.class.send(:finalizer, @ptr)
      ObjectSpace.define_finalizer(self, @finalizer)
    end

    # -- lifecycle --------------------------------------------------------
    def close
      return if @ptr.nil?

      ObjectSpace.undefine_finalizer(self)
      Lib.rb_dsp_free(@ptr)
      @ptr = nil
    end

    def closed?
      @ptr.nil?
    end

    def self.finalizer(ptr)
      proc { Lib.rb_dsp_free(ptr) }
    end
    private_class_method :finalizer

    # -- configuration ----------------------------------------------------
    def set_input_frequency(hz)
      Lib.rb_dsp_set_input_frequency(@ptr, Integer(hz))
    end

    def flush
      Lib.rb_dsp_flush(@ptr)
    end

    def eq_enable(enable)
      Lib.rb_dsp_eq_enable(@ptr, RockboxFFI.b(enable))
    end

    # Configure one EQ band (0..=9). Band 0 low shelf, 9 high shelf.
    def set_eq_band(band, cutoff_hz, q, gain_db)
      Lib.rb_dsp_set_eq_band(@ptr, Integer(band), Integer(cutoff_hz), Float(q), Float(gain_db))
    end

    def set_eq_precut(db)
      Lib.rb_dsp_set_eq_precut(@ptr, Float(db))
    end

    def set_tone(bass_db, treble_db)
      Lib.rb_dsp_set_tone(@ptr, Integer(bass_db), Integer(treble_db))
    end

    def set_tone_cutoffs(bass_hz, treble_hz)
      Lib.rb_dsp_set_tone_cutoffs(@ptr, Integer(bass_hz), Integer(treble_hz))
    end

    def set_surround(delay_ms, balance, fx1, fx2)
      Lib.rb_dsp_set_surround(@ptr, Integer(delay_ms), Integer(balance), Integer(fx1), Integer(fx2))
    end

    def set_channel_config(mode)
      Lib.rb_dsp_set_channel_config(@ptr, Integer(mode))
    end

    def set_stereo_width(percent)
      Lib.rb_dsp_set_stereo_width(@ptr, Integer(percent))
    end

    def set_compressor(threshold, makeup_gain, ratio, knee, release_time, attack_time)
      Lib.rb_dsp_set_compressor(
        @ptr, Integer(threshold), Integer(makeup_gain), Integer(ratio),
        Integer(knee), Integer(release_time), Integer(attack_time)
      )
    end

    # mode: see DspReplayGainMode (TRACK=0, ALBUM=1, SHUFFLE=2, OFF=3).
    def set_replaygain(mode, noclip, preamp_db)
      Lib.rb_dsp_set_replaygain(@ptr, Integer(mode), RockboxFFI.b(noclip), Float(preamp_db))
    end

    # Per-track gains in plain dB / peaks as linear amplitude (1.0 = full
    # scale). nil for any absent tag (mapped to the ABI's NaN sentinel).
    def set_replaygain_gains(track_gain_db: nil, album_gain_db: nil, track_peak: nil, album_peak: nil)
      Lib.rb_dsp_set_replaygain_gains(
        @ptr, opt(track_gain_db), opt(album_gain_db), opt(track_peak), opt(album_peak)
      )
    end

    # Native Q7.24 linear factors (the raw_* fields from Metadata.read),
    # 0 = not tagged.
    def set_replaygain_gains_raw(track_gain, album_gain, track_peak, album_peak)
      Lib.rb_dsp_set_replaygain_gains_raw(
        @ptr, Integer(track_gain), Integer(album_gain), Integer(track_peak), Integer(album_peak)
      )
    end

    # -- processing -------------------------------------------------------
    # Run interleaved stereo S16 +samples+ (an Array of Integers) through the
    # pipeline. Returns a new Array of processed samples (length may differ
    # from the input — the resampler buffers).
    def process(samples)
      raise ArgumentError, "input must be interleaved stereo (even length)" if samples.length.odd?

      input = samples.pack("s<*")               # int16 little-endian
      out_len = "\0" * Fiddle::SIZEOF_VOIDP      # size_t* out-param (written in place)
      out_ptr = Lib.rb_dsp_process(@ptr, input, samples.length, out_len)

      produced = out_len.unpack1(Fiddle::SIZEOF_VOIDP == 8 ? "Q" : "L")
      return [] if out_ptr.null? || produced.zero?

      begin
        out_ptr.size = produced * 2
        out_ptr[0, produced * 2].unpack("s<*")
      ensure
        Lib.rb_buffer_free(out_ptr, produced)
      end
    end

    private

    NAN = Float::NAN
    private_constant :NAN

    # nil -> NaN, the ABI's "tag absent" sentinel.
    def opt(value)
      value.nil? ? NAN : Float(value)
    end
  end

  # Generate +seconds+ of a sine as interleaved stereo int16 (an Array).
  def self.sine_stereo(freq_hz, seconds, rate, amplitude = 16_000)
    n = (seconds * rate).to_i
    buf = Array.new(n * 2)
    (0...n).each do |i|
      s = (Math.sin(i * 2.0 * Math::PI * freq_hz / rate) * amplitude).to_i
      s = -32_768 if s < -32_768
      s = 32_767 if s > 32_767
      buf[i * 2] = s
      buf[(i * 2) + 1] = s
    end
    buf
  end
end
